### PR TITLE
[AIRFLOW-2066] Add operator to create an empty BigQuery table

### PR DESCRIPTION
**Pull request by kaxil**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2066

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Add operator that creates a new, empty table in the specified BigQuery dataset, optionally with schema.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Added `BigQueryCreateEmptyTableOperatorTest`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by kaxil** [comment]
cc @Fokko @fenglu-g @criccomini @jgao54

File: airflow/contrib/hooks/bigquery_hook.py [hunk 1]
```diff
@@ -207,6 +207,88 @@ def __init__(self, service, project_id, use_legacy_sql=True):
...
+        Creates a new, empty table in the dataset.
+
+        :param project_id: The project to create the table into.
+        :type project_id: str
+        :param dataset_id: The dataset to create the table into.
+        :type dataset_id: str
+        :param table_id: The Name of the table to be created.
+        :type table_id: str
+        :param schema_fields: The schema field list as defined here: ::
+            https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#resource
```
Comment (fenglu-db) [inline airflow/contrib/hooks/bigquery_hook.py]: could you also provide an example here for schema_fields?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 2]
```diff
@@ -207,6 +207,88 @@ def __init__(self, service, project_id, use_legacy_sql=True):
...
+                          pageToken=tables_list_resp['nextPageToken'])\
+                    .execute()
+            # If there is no next page, then the table doesn't exist.
+            else:
+                # do insert
+                self.log.info('Creating Table %s:%s.%s',
+                              project_id, dataset_id, table_id)
+
+                try:
+                    self.service.tables().insert(
```
Comment (fenglu-db) [inline airflow/contrib/hooks/bigquery_hook.py]: what if you don't list before insert? trying to understand the value added piece of checking existence as opposed to blind insertion.

File: airflow/contrib/hooks/bigquery_hook.py [hunk 3]
```diff
@@ -207,6 +207,88 @@ def __init__(self, service, project_id, use_legacy_sql=True):
...
+            https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#resource
+        :type schema_fields: list
+        :param time_partitioning: configure optional time partitioning fields i.e.
+            partition by field, type and expiration as per API specifications.
+
+            .. seealso::
+            https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#timePartitioning
+        :type time_partitioning: dict
+
+        :return:
```
Comment (fenglu-db) [inline airflow/contrib/hooks/bigquery_hook.py:242]: return bool? or eliminate the return True you have?

File: airflow/contrib/operators/bigquery_operator.py [hunk 4]
```diff
@@ -121,3 +124,111 @@ def on_kill(self):
...
+    :param dataset_id: The dataset to create the table into.
+    :type dataset_id: string
+    :param table_id: The Name of the table to be created.
+    :type table_id: string
+    :param schema_fields: If set, the schema field list as defined here:
+        https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.load
+    :type schema_fields: list
+    :param schema_object: If set, a GCS object path pointing to a .json file that
+        contains the schema for the table.
+    :type schema_object: string
```
Comment (fenglu-db) [inline airflow/contrib/operators/bigquery_operator.py]: do you want to merge schema_object and schema_object_gcs_bucket as gcs_schema_object?

File: airflow/contrib/operators/bigquery_operator.py [hunk 5]
```diff
@@ -121,3 +124,111 @@ def on_kill(self):
...
+
+    :param project_id: The project to create the table into.
+    :type project_id: string
+    :param dataset_id: The dataset to create the table into.
+    :type dataset_id: string
+    :param table_id: The Name of the table to be created.
+    :type table_id: string
+    :param schema_fields: If set, the schema field list as defined here:
+        https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.load
+    :type schema_fields: list
```
Comment (fenglu-db) [inline airflow/contrib/operators/bigquery_operator.py]: nit: example on schema_fields.

**Review by fenglu-db — COMMENTED** [review]
Some minor changes

File: airflow/contrib/hooks/bigquery_hook.py (hunk 1 above)
Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py]: Example Added

File: airflow/contrib/hooks/bigquery_hook.py (hunk 3 above)
Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py:242]: eliminated the return True

Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py:242]: eliminated the return True

**Comment by kaxil** [comment]
@fenglu-g Thank you for the time to review it. I have made the changes you had suggested.

File: tests/contrib/operators/test_bigquery_operator.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    except ImportError:
+        mock = None
+
+TASK_ID = 'test-bq-create-empty-table-operator'
+TEST_DATASET = 'test-dataset'
+TEST_PROJECT_ID = 'test-project'
+TEST_TABLE_ID = 'test-table-id'
+
+
+class BigQueryCreateEmptyTableOperatorTest(unittest.TestCase):
```
Comment (fenglu-db) [inline tests/contrib/operators/test_bigquery_operator.py:33]: nit: add unit test that exercise _parse_gcs_url?

**Review by fenglu-db — APPROVED** [review]
LGTM Thank you @kaxil for help improving GCP operators!

File: tests/contrib/operators/test_bigquery_operator.py (hunk 6 above)
Comment (kaxil) [reply tests/contrib/operators/test_bigquery_operator.py:33]: Have added test here: https://github.com/apache/incubator-airflow/pull/3006/files#diff-bdc2f52d45b60d6afb163346efc72774

**Comment by kaxil** [comment]
Welcome @fenglu-g. I have added the test for `_parse_gcs_url ` [here](https://github.com/apache/incubator-airflow/pull/3006/files#diff-bdc2f52d45b60d6afb163346efc72774). Please take a look.

**Pull request closed without merging**